src/cargo/ops/cargo_compile/packages.rs

```diff
@@ -189,19 +209,20 @@ fn emit_pattern_not_found(
 
 /// Given a list opt-in or opt-out package selection strings, generates two
 /// collections that represent glob patterns and package names respectively.
-fn opt_patterns_and_names(
+fn opt_patterns_and_ids(
     opt: &[String],
-) -> CargoResult<(Vec<(glob::Pattern, bool)>, BTreeSet<&str>)> {
+) -> CargoResult<(Vec<(glob::Pattern, bool)>, BTreeSet<PackageIdSpec>)> {
     let mut opt_patterns = Vec::new();
-    let mut opt_names = BTreeSet::new();
+    let mut opt_ids = BTreeSet::new();
     for x in opt.iter() {
         if PackageIdSpec::parse(x).is_err() && is_glob_pattern(x) {
             opt_patterns.push((build_glob(x)?, false));
         } else {
-            opt_names.insert(String::as_str(x));
+            let id = PackageIdSpec::parse(x)?;
+            opt_ids.insert(id);
         }
```

Comment:
Something like this to avoid parsing twice?

```suggestion
        match PackageIdSpec::parse(x) {
            Ok(spec) => {
                opt_ids.insert(spec);
            }
            Err(_) if is_glob_pattern(x) => opt_patterns.push((build_glob(x)?, false)),
            Err(e) => return Err(e.into()),
        }
```

Reply:
Good suggestion!